### PR TITLE
feat: codebase metrics calculations

### PR DESCRIPTION
**Pull request by vishalshenoy**
Eventually convert utils into dashboard for codebase metrics - this is demos for relevant code metrics

### CODE REVIEW

**Comment by vishalshenoy** [comment]
Cyclomatic Complexity, Depth of Inheritance, Halstead Volume, Line stats, Maintainability Index all calculated with codegen-sdk

**Review by kopekC — APPROVED** [review]

**Comment by codegen-team** [comment]
CodegenBot is starting to review the PR please wait...

File: codegen-examples/examples/modal_repo_analytics/utils/depth_of_inherritance.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
+import codegen
+from codegen import Codebase
+
+
+def calculate_doi(cls):
+    """Calculate the depth of inheritance for a given class."""
+    return len(cls.superclasses)
```
Comment (codegen-team) [inline codegen-examples/examples/modal_repo_analytics/utils/depth_of_inherritance.py:7]: Consider adding threshold recommendations for inheritance depth. Generally, a depth > 3 might indicate potential design issues. Also, consider tracking and reporting the full inheritance chain, not just the count.

File: codegen-examples/examples/modal_repo_analytics/utils/cyclomatic_complexity.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+            complexity += len(getattr(statement, "except_blocks", []))
+
+        if hasattr(statement, "condition") and isinstance(statement.condition, str):
+            complexity += statement.condition.count(" and ") + statement.condition.count(" or ")
+
+        if hasattr(statement, "nested_code_blocks"):
+            for block in statement.nested_code_blocks:
+                complexity += analyze_block(block)
+
+        return complexity
```
Comment (codegen-team) [inline codegen-examples/examples/modal_repo_analytics/utils/cyclomatic_complexity.py:42]: The complexity calculation could be enhanced by considering boolean operator short-circuiting (&&, ||) in conditions, as these create additional paths through the code. Also, consider adding function-level recommendations for when to split complex functions.

File: codegen-examples/examples/modal_repo_analytics/utils/lines.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        line = lines[i]
+        code_part = line
+        if not in_multiline and "#" in line:
+            comment_start = line.find("#")
+            if not re.search(r'["\'].*#.*["\']', line[:comment_start]):
+                code_part = line[:comment_start].strip()
+                if line[comment_start:].strip():
+                    comments += 1
+
+        # Handle multi-line strings/comments
```
Comment (codegen-team) [inline codegen-examples/examples/modal_repo_analytics/utils/lines.py:37]: The line counting logic could be improved by handling more edge cases:
1. Inline strings containing comment characters
2. Triple-quoted strings that span multiple lines
3. Nested multi-line strings

Consider using the tokenize module from the standard library to handle these cases more robustly.

**Comment by codegen-team** [comment]
Overall Implementation Review:

Strengths:
1. Well-structured code with clear separation of concerns
2. Good use of type hints and docstrings
3. Comprehensive implementation of various code metrics
4. Nice output formatting with emojis for better readability
5. Good error handling and edge case consideration

Suggestions for Enhancement:
1. Consider adding configuration options for thresholds and reporting formats
2. Add unit tests for edge cases in each metric calculation
3. Consider implementing caching for performance optimization when analyzing large codebases
4. Add aggregate reporting across multiple files/modules
5. Consider adding trend analysis capabilities for tracking metrics over time

The code is production-ready but could benefit from these enhancements for better maintainability and usability.

**Comment by codegen-team** [comment]
CodegenBot is starting to review the PR please wait...

**Comment by codegen-team** [comment]
### Overall Assessment

This PR introduces a comprehensive set of code analysis utilities that will be valuable for repository analytics. The implementation is generally well-structured and includes important metrics used in software engineering.

### Strengths
- Comprehensive coverage of key metrics
- Good documentation and clear function purposes
- Consistent error handling and edge cases
- Well-organized code structure

### Areas for Improvement

1. **Code Duplication**: There's significant duplication in operator/operand analysis between `halstead_volume.py` and `maintainability_index.py`. Consider extracting this into a shared utility module.

2. **Error Handling**: While basic error handling exists, some edge cases might need more attention:
   - Division by zero protection in maintainability calculations
   - Handling of malformed/invalid code structures
   - Proper error propagation

3. **Performance Considerations**: For large codebases, some operations might be expensive:
   - Consider adding caching mechanisms for repeated calculations
   - Look into batch processing for file analysis
   - Add progress indicators for long-running operations

4. **Testing**: Would be beneficial to add:
   - Unit tests for each metric calculation
   - Integration tests with known codebase examples
   - Edge case testing for various code structures

### Suggestions for Future Enhancements
1. Add configuration options for metric thresholds
2. Implement reporting formats (JSON, CSV, HTML)
3. Consider adding trend analysis capabilities
4. Add support for different programming languages

File: codegen-examples/examples/modal_repo_analytics/utils/halstead_volume.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+                    operators.extend([op.source for op in expr.operators])
+                    operands.extend([elem.source for elem in expr.elements])
+                elif isinstance(expr, UnaryExpression):
+                    operators.append(expr.ts_node.type)
+                    operands.append(expr.argument.source)
+                elif isinstance(expr, ComparisonExpression):
+                    operators.extend([op.source for op in expr.operators])
+                    operands.extend([elem.source for elem in expr.elements])
+
+        if hasattr(statement, "expression"):
```
Comment (codegen-team) [inline codegen-examples/examples/modal_repo_analytics/utils/halstead_volume.py:31]: Consider extracting this operator/operand analysis into a shared utility module to avoid duplication with maintainability_index.py. This would improve maintainability and ensure consistent behavior across metrics.

File: codegen-examples/examples/modal_repo_analytics/utils/depth_of_inherritance.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
+import codegen
+from codegen import Codebase
+
+
+def calculate_doi(cls):
```
Comment (codegen-team) [inline codegen-examples/examples/modal_repo_analytics/utils/depth_of_inherritance.py:5]: The depth of inheritance calculation could be enhanced to handle multiple inheritance paths and provide the maximum depth across all paths. Currently, it only uses len(cls.superclasses) which might not capture the full inheritance hierarchy.

File: codegen-examples/examples/modal_repo_analytics/utils/cyclomatic_complexity.py [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+from codegen.sdk.core.statements.for_loop_statement import ForLoopStatement
+from codegen.sdk.core.statements.if_block_statement import IfBlockStatement
+from codegen.sdk.core.statements.try_catch_statement import TryCatchStatement
+from codegen.sdk.core.statements.while_statement import WhileStatement
+
+
+def cc_rank(complexity):
+    if complexity < 0:
+        raise ValueError("Complexity must be a non-negative value")
+
```
Comment (codegen-team) [inline codegen-examples/examples/modal_repo_analytics/utils/cyclomatic_complexity.py:12]: Consider adding a configuration parameter for the complexity thresholds. Currently, the ranks are hardcoded, but different projects might have different standards for what constitutes acceptable complexity.

File: codegen-examples/examples/modal_repo_analytics/utils/lines.py [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+        if ('"""' in line or "'''" in line) and not (line.count('"""') % 2 == 0 or line.count("'''") % 2 == 0):
+            if in_multiline:
```
Comment (codegen-team) [inline codegen-examples/examples/modal_repo_analytics/utils/lines.py:39]: The line counting logic could benefit from more robust handling of edge cases, particularly around multi-line strings and comments. Consider using the tokenize module from Python's standard library to handle these cases more reliably.

File: codegen-examples/examples/modal_repo_analytics/utils/maintainability_index.py [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+    except (ValueError, TypeError):
+        return 0
+
+
+def get_maintainability_rank(mi_score: float) -> str:
+    """Convert maintainability index score to a letter grade.
+
+    Args:
+        mi_score: The maintainability index score
+
```
Comment (codegen-team) [inline codegen-examples/examples/modal_repo_analytics/utils/maintainability_index.py:141]: The maintainability index calculation could benefit from some performance optimizations. Consider caching the results of expensive calculations (like Halstead volume) and reusing them across different metrics. Also, consider adding progress reporting for large codebases.
